ibis/pandas/udf.py

```diff
@@ -216,30 +216,47 @@ def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
     input_type = op.input_type
     nargs = len(input_type)
 
-    # An execution rule for a simple aggregate node
+    # An execution rule for ungrouped aggregations. Handles:
+    #   1) AggregationNode with no grouping
+    #   2) Some aggregation over a window with no grouping
     @execute_node.register(type(op), *(itertools.repeat(pd.Series, nargs)))
     def execute_udaf_node(op, *args, **kwargs):
-        return op.func(*args)
+        aggcontext = kwargs.pop('aggcontext', None)
+        assert aggcontext is not None, 'aggcontext is None'
 
-    # An execution rule for a grouped aggregation node. This
-    # includes aggregates applied over a window.
+        func = op.func
+
+        if isinstance(aggcontext, Window):
+            # If aggcontext is a Window, then we must be aggregating over
+            # a bounded window, so we need aggcontext.agg to handle the
+            # rolling/expanding window logic.
+            return aggcontext.agg(args[0], func, *args, **kwargs)
+        else:
```

Comment:
Can we assert exactly the type of aggcontext here? e.g

```
if isinstance(aggcontext, Window):

elif isinstance(aggcontext, ???):

else:
    raise AssertionError(f"Unexpected aggcontext of type {type{asggcontext}")

```

Reply:
Done

Reply:
(Ended up removing this `AssertionError` in a later commit so that we continue to support custom `AggregationContexts`)